tests/integration/index/query_with_relation_filter_test.go

```diff
@@ -362,7 +362,7 @@ func TestQueryWithIndexOnOneToOnePrimaryRelation_IfFilterOnIndexedFieldOfRelatio
 				// we make 3 index fetch to get the 3 address with city == "Montreal"
 				// then we scan all 10 users to find one with matching "address_id" for each address
 				// after this we fetch the name of each user
-				Asserter: testUtils.NewExplainAsserter().WithFieldFetches(33).WithIndexFetches(3),
+				Asserter: testUtils.NewExplainAsserter().WithFieldFetches(60).WithIndexFetches(3),
```

Comment:
question: Whats the context on field fetches increases, is it because all fields are read now?

Reply:
Noted in the description.  It is a large increase due to the lack of indexing (and other potential optimisations), so we end up doing a full scan for every parent record - field count here has gone from (3*10*1+(1*3)) to (3*10*2) (note if you rename `address` to `baddress` on develop branch you get the same result).

Reply:
I missed the part in the description explaining why number of fields fetched increased.

> Previously, filtering would be done as soon as all the fields for the filter had been read into memory. Now filtering is done once all selected fields have been read.

Is it this one?

And what do you mean by "lack of indexing"?

I still don't understand why it fetches 6 fields for every User doc.

Reply:
yes, you mentioned "lack of indexing" and I asked you afterwards what you meant. You didn't answer yet.

It's clear why we do full scan on primary docs for every related doc, but it's not clear why you need to fetch 2 fields for every primary doc. The request contains only `name` in selected fields.
What it did before: for every primary doc we would read `address_id` to compare it with related doc's we found using index. That's already 30: 10 times for 3 docs. Then for every matching primary doc we would read additionally `name`. So +3.

I assume the answer lies somewhere here "Previously, filtering would be done as soon as all the fields for the filter had been read into memory. Now filtering is done once all selected fields have been read."
Which for this context I interpret as the following: while doing full scan we always fetch 2 fields: `address_id` to match against the related doc and `name` because it's in the selection set. In this case I wonder if we really have to read the selected fields before filter.

Sorry, I'm just trying to understand the new behaviour. Not trying to be picky.

Reply:
> You didn't answer yet.

Ah I thought you missed my earlier (small expansion) :)  I'll expand further.

Currently a full scan is done for every parent.  This is not ideal.  Indexing the child to parent would remove the need for this, and I thought we had already discussed as team doing that - this would convert the full table scan into a point lookup.  Atm I think this can be done already by users by explicitly declaring the index from secondary side.

There also are other non-index optimizations we can do here too long term that would at least allow us to only scan child records once for all parents (instead of a full scan for every parent) if we dont have an index.

> but it's not clear why you need to fetch 2 fields for every primary doc

Two fields are required.  One for the filter, one for the select.  `develop` branch does this anyway - if you rename `address` to `oddress` (or anything lexographically larger than `name`) it will fetch `60` fields.  The `33` field count in `develop` is also something of a lie atm anyway, `develop` currently reads `60` fields (value and key) from disk anyway, no io is saved, it just skips a small amount of Defra Go code for the other `27` fields that are discarded.

Once we integrate the CoreKV lib into Defra, with it's exposed `seek` logic we can actually rework the many performance deficiencies of the way the fetcher field stuff works (old and current, they both work in roughly the same horrible way here).  

Once CoreKV is integrated, and the other nastiness reworked, we can actually measure the effect of skipping the extra io of fields lexicographically larger than the largest filter field - this is where John/Fred disagree with me RE performance - IO can actually be really quite cheap, and I think there are reasonable odds of the logic required to skip those properties having either a negative or minimal effect on overall system performance.  Once we have measured the performance of the two (simple vs seek) we can make a choice as to whether we want to take on the maintenance cost of the seek-based solution.

Reply:
okay, I see. Thanks for detailed explanation. Was very helpful.